src/lib/pubkey/kyber/kyber_common/kyber.cpp

```diff
@@ -441,38 +442,13 @@ class Polynomial {
        * Run rejection sampling on uniform random bytes to generate uniform
        * random integers mod q.
        */
-      static Polynomial sample_rej_uniform(Kyber_XOF& xof) {
-         class XOF_Reader {
-            public:
-               XOF_Reader(Kyber_XOF& xof) : m_xof(xof), m_buf(), m_buf_pos(m_buf.size()) {}
-
-               std::array<uint8_t, 3> next_3_bytes() {
-                  if(m_buf_pos == m_buf.size()) {
-                     m_xof.write_output(m_buf);
-                     m_buf_pos = 0;
-                  }
-
-                  BOTAN_ASSERT_NOMSG(m_buf_pos + 3 <= m_buf.size());
-
-                  std::array<uint8_t, 3> buf{m_buf[m_buf_pos], m_buf[m_buf_pos + 1], m_buf[m_buf_pos + 2]};
-
-                  m_buf_pos += 3;
-
-                  return buf;
-               }
-
-            private:
-               Kyber_XOF& m_xof;
-               std::array<uint8_t, 3 * 64> m_buf;
-               size_t m_buf_pos;
-         };
-
-         auto xof_reader = XOF_Reader(xof);
+      static Polynomial sample_rej_uniform(XOF& xof) {
          Polynomial p;
 
          size_t count = 0;
          while(count < p.size()) {
-            auto buf = xof_reader.next_3_bytes();
+            std::array<uint8_t, 3> buf;
+            xof.output(buf);
```

Comment:
Does not buffering anymore cause a performance regression? IIRC this change had a pretty noticeable impact

Reply:
For keygen it does, indeed. I'll look into it.

|                   | keygen/s (master) | keygen/s (XOF) |
| ----------------- | ----------------- | -------------- |
| Kyber-512-r3      | 16095             | 15587          |
| Kyber-512-90s-r3  | 27645             | 14944          |
| Kyber-768-r3      | 13160             | 15992          |
| Kyber-768-90s-r3  | 26266             | 14622          |
| Kyber-1024-r3     | 12685             | 12856          |
| Kyber-1024-90s-r3 | 16779             | 7289           |

Reply:
The "XOF w/ buffer" uses the following patch. I didn't push that, as it adds an allocation and doesn't really speed up the SHA-3 mode.

```diff
diff --git a/src/lib/pubkey/kyber/kyber_common/kyber.cpp b/src/lib/pubkey/kyber/kyber_common/kyber.cpp
index 107e93f28..fa43c2055 100644
@@ -445,10 +445,12 @@ class Polynomial {
       static Polynomial sample_rej_uniform(XOF& xof) {
          Polynomial p;

+         const auto xof_output = xof.output<secure_vector<uint8_t>>(p.size() * 3);
+         BufferSlicer r(xof_output);
+
          size_t count = 0;
          while(count < p.size()) {
-            std::array<uint8_t, 3> buf;
-            xof.output(buf);
+            const auto buf = r.take(3);

             const uint16_t val0 = ((buf[0] >> 0) | (static_cast<uint16_t>(buf[1]) << 8)) & 0xFFF;
             const uint16_t val1 = ((buf[1] >> 4) | (static_cast<uint16_t>(buf[2]) << 4)) & 0xFFF;
```

Reply:
OK if the regression is only for 90s mode, and only for keygen, seems fine to leave as is.